### PR TITLE
Support ExtendedSecret certs signed by the Kube API Server

### PR DESCRIPTION
**Pull request by edwardstudy**
[#167555390](https://www.pivotaltracker.com/story/show/167555390)

- Added two options to support CSR for ExtendedSecret:
```
SignerType       SignerType        `json:"signerType,omitempty"`
Usages           []certv1.KeyUsage `json:"usages,omitempty"`
```

- Added CSR controller to watch CSR and approve it and then persists certificate signed by Kube API Server

Verification steps: 
```
$ k apply -f ./docs/examples/extended-secret/certificate.yaml

$ k get secret
NAME                                   TYPE                                  DATA   AGE
...
nats-deployment.var-nats-cert    # This cert signed by API Server
```

### CODE REVIEW

File: pkg/kube/controllers/extendedsecret/certificatesigningrequest_reconciler.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+
+		err = r.deleteSecret(ctx, privatekeySecret)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		return reconcile.Result{}, nil
+	}
+
+	err = r.approveRequest(ctx, instance.Name)
```
Comment (aduffeck) [inline pkg/kube/controllers/extendedsecret/certificatesigningrequest_reconciler.go]: I think this should better go into an `else` branch.

File: pkg/kube/util/names/names.go [hunk 2]
```diff
@@ -180,6 +180,11 @@ func OrdinalFromPodName(name string) int {
 	return podOrdinal
 }
 
+// PrivateKeySecretName returns a Secret name for a given ExtendedSecret
+func PrivateKeySecretName(extendedSecretName string) string {
```
Comment (aduffeck) [inline pkg/kube/util/names/names.go]: Could we make this name a bit more precise, i.e. ` CsrPrivateKeySecretName`?

File: pkg/kube/controllers/extendedsecret/certificatesigningrequest_reconciler.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+			ctxlog.Info(ctx, "Skip reconcile: certificateSigningRequest not found")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		ctxlog.Info(ctx, "Error reading the object")
+		return reconcile.Result{}, errors.Wrap(err, "Error reading certificateSigningRequest")
+	}
+
+	if len(instance.Status.Certificate) != 0 {
+		annotations := instance.GetAnnotations()
```
Comment (aduffeck) [inline pkg/kube/controllers/extendedsecret/certificatesigningrequest_reconciler.go]: I'm just wondering, why do we need these annotations? Couldn't we get the secret name and namespace from the owner object (ExtendedSecret)?

File: pkg/kube/controllers/extendedsecret/certificatesigningrequest_reconciler.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+		}
+		namespace, ok := annotations[esv1.AnnotationExSecretNamespace]
+		if !ok {
+			ctxlog.WithEvent(instance, "NotFoundError").Errorf(ctx, "failed to lookup exSecret namespace from certificateSigningRequest '%s' annotations", instance.Name)
+			return reconcile.Result{}, nil
+		}
+
+		privatekeySecret, err := r.getSecret(ctx, namespace, names.PrivateKeySecretName(instance.Name))
+		if err != nil {
+			return reconcile.Result{}, err
```
Comment (aduffeck) [inline pkg/kube/controllers/extendedsecret/certificatesigningrequest_reconciler.go]: How about adding a log message here?

File: pkg/kube/controllers/extendedsecret/certificatesigningrequest_reconciler.go (hunk 1 above)
Comment (aduffeck) [reply pkg/kube/controllers/extendedsecret/certificatesigningrequest_reconciler.go]: And couldn't we pass the object directly instead of its name to get rid of another get request?

File: pkg/kube/controllers/extendedsecret/extendedsecret_reconciler.go [hunk 5]
```diff
@@ -124,6 +127,7 @@ func (r *ReconcileExtendedSecret) Reconcile(request reconcile.Request) (reconcil
 func (r *ReconcileExtendedSecret) updateExSecret(ctx context.Context, instance *esv1.ExtendedSecret) error {
 	obj := instance.DeepCopy()
 	op, err := controllerutil.CreateOrUpdate(ctx, r.client, obj, func() error {
+		obj.Spec = instance.Spec
```
Comment (aduffeck) [inline pkg/kube/controllers/extendedsecret/extendedsecret_reconciler.go]: Why is this change needed?

File: pkg/kube/controllers/extendedsecret/certificatesigningrequest_reconciler.go (hunk 3 above)
Comment (edwardstudy) [reply pkg/kube/controllers/extendedsecret/certificatesigningrequest_reconciler.go]: The truth is that:
1. CSR is no-namespaced
2. the owner and owned object are both namespace-scoped and in the same namespace

So getting the owner's namespace from CSR could not get right namespace. 
My idea is to set the owner's namespace to annotations.

File: pkg/kube/controllers/extendedsecret/certificatesigningrequest_reconciler.go (hunk 1 above)
Comment (edwardstudy) [reply pkg/kube/controllers/extendedsecret/certificatesigningrequest_reconciler.go]: Oh, it's helpful for reducing update conflicts. 
And `r.client` could not update subresource `approval`. we have to use `client-go.client` to update CSR approval directly: 
https://github.com/kubernetes-sigs/controller-runtime/issues/172

File: pkg/kube/controllers/extendedsecret/extendedsecret_reconciler.go (hunk 5 above)
Comment (edwardstudy) [reply pkg/kube/controllers/extendedsecret/extendedsecret_reconciler.go]: It's helpful to update default `spec.request.signerType` if it's certificate.

File: docs/examples/extended-secret/certificate.yaml [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        Name: cert-ca
+      CARef:
+        Key: certificate
+        Name: cert-ca
+      alternativeNames:
+        - foo.com
+        - '*.foo.com'
+      commonName: routerSSL
+      isCA: false
+      signerType: external
```
Comment (viovanov) [inline docs/examples/extended-secret/certificate.yaml]: This looks like it's not using the API server to sign

File: docs/controllers/extendedsecret.md [hunk 7]
```diff
@@ -32,6 +33,24 @@ A pluggable implementation for generating certificates and passwords.
 
 The developer can specify policies for rotation (e.g. automatic or not) and how secrets are created (e.g. password complexity, certificate expiration date, etc.).
 
+### Auto-approving Certificates
+
+A certificate `ExtendedSecret` can be signed by the Kube API Server. The ExtendedSecret Controller is be responsible for generating certificate signing request and creating CertificateSigningRequest:
```
Comment (viovanov) [inline docs/controllers/extendedsecret.md]: For generating ... And _approving_

File: pkg/credsgen/in_memory_generator/certificate.go [hunk 8]
```diff
@@ -38,6 +38,27 @@ func (g InMemoryGenerator) GenerateCertificate(name string, request credsgen.Cer
...
+// GenerateCertificateSigningRequest Generates a certificate signing request and private key
+func (g InMemoryGenerator) GenerateCertificateSigningRequest(request credsgen.CertificateGenerationRequest) ([]byte, []byte, error) {
+	cfssllog.Level = cfssllog.LevelWarning
+
+	var csReq, privateKey []byte
+
+	// Generate certificate request
+	certReq := &csr.CertificateRequest{KeyRequest: &csr.BasicKeyRequest{A: g.Algorithm, S: g.Bits}}
+
+	certReq.Hosts = append(certReq.Hosts, request.CommonName)
```
Comment (viovanov) [inline pkg/credsgen/in_memory_generator/certificate.go]: Do you know if it works with IPs?

File: pkg/kube/apis/extendedsecret/v1alpha1/types.go [hunk 9]
```diff
@@ -12,20 +13,33 @@ import (
...
+	Certificate SecretType = "certificate"
+	SSHKey      SecretType = "ssh"
+	RSAKey      SecretType = "rsa"
+)
+
+// SignerType defines the type of the certificate signer
+type SignerType = string
+
+// Valid values for signer types
+const (
```
Comment (viovanov) [inline pkg/kube/apis/extendedsecret/v1alpha1/types.go]: What do these mean?

Comment (viovanov) [reply pkg/kube/apis/extendedsecret/v1alpha1/types.go]: Can you please document?

File: docs/examples/extended-secret/certificate.yaml (hunk 6 above)
Comment (edwardstudy) [reply docs/examples/extended-secret/certificate.yaml]: Removed CA infos. Sorry for misleading example.

File: pkg/kube/apis/extendedsecret/v1alpha1/types.go (hunk 9 above)
Comment (edwardstudy) [reply pkg/kube/apis/extendedsecret/v1alpha1/types.go]: Done

File: pkg/credsgen/in_memory_generator/certificate.go (hunk 8 above)
Comment (edwardstudy) [reply pkg/credsgen/in_memory_generator/certificate.go]: It should be. thinking a case about it.

File: pkg/kube/controllers/extendedsecret/certificatesigningrequest_reconciler.go (hunk 4 above)
Comment (edwardstudy) [reply pkg/kube/controllers/extendedsecret/certificatesigningrequest_reconciler.go]: done

File: pkg/kube/controllers/extendedsecret/extendedsecret_reconciler.go (hunk 5 above)
Comment (edwardstudy) [reply pkg/kube/controllers/extendedsecret/extendedsecret_reconciler.go]: refactored CreateOrUpdate mutateFn

**Comment by edwardstudy** [comment]
operator can consume following CR and create cert signed by API server
```
---
apiVersion: fissile.cloudfoundry.org/v1alpha1
kind: ExtendedSecret
metadata:
  name: generate-certificate
spec:
  request:
    certificate:
      alternativeNames:
        - foo.com
        - '*.foo.com'
      commonName: routerSSL
      signerType: cluster
      usages: ["server auth", "client auth"]
  secretName: gen-certificate
  type: certificate
```

File: e2e/kube/examples_test.go [hunk 10]
```diff
@@ -170,33 +170,33 @@ var _ = Describe("Examples", func() {
...
+					Expect(err).ToNot(HaveOccurred(), "error restarting cf-operator")
 
 					By("Checking for pods not created")
 					err = kubectlHelper.Wait(namespace, "ready", "pod/nats-deployment-nats-v2-0", 10*time.Second)
 					Expect(err).To(HaveOccurred(), "error unexpected version of instance group is created")
 
 					By("Checking for secrets not created")
 					exist, err := kubectlHelper.SecretExists(namespace, "nats-deployment.bpm.nats-v2")
-					Expect(err).To(HaveOccurred())
+					Expect(err).ToNot(HaveOccurred(), "error getting secret/nats-deployment.bpm.nats-v2")
```
Comment (aduffeck) [inline e2e/kube/examples_test.go]: Why was the expectation negated here and in the cases below?

Comment (edwardstudy) [reply e2e/kube/examples_test.go]: This case is that we don't want new versions to be created when the operator restart. I changed `kubectlHelper.SecretExists` that doesn't return an error when object not found because I want to use it in the WaitForSecret func that didn't break poll when object not found.

Comment (aduffeck) [reply e2e/kube/examples_test.go]: Got it.
